v-next/hardhat/src/internal/builtin-plugins/coverage/coverage-manager.ts

```diff
@@ -0,0 +1,106 @@
+import type { CoverageHits, CoverageManager } from "./types.js";
+import type { EdrProvider } from "../network-manager/edr/edr-provider.js";
+
+import { randomUUID } from "node:crypto";
+import path from "node:path";
+
+import { assertHardhatInvariant } from "@nomicfoundation/hardhat-errors";
+import {
+  ensureDir,
+  getAllFilesMatching,
+  readJsonFile,
+  remove,
+  writeJsonFile,
+} from "@nomicfoundation/hardhat-utils/fs";
+
+let coverageManager: CoverageManager | undefined;
+
+export function getOrCreateCoverageManager(): CoverageManager {
+  if (coverageManager === undefined) {
+    coverageManager = new CoverageManagerImplementation();
+  }
+  return coverageManager;
+}
+
+export class CoverageManagerImplementation implements CoverageManager {
+  readonly #providers: Record<string, EdrProvider> = {};
+  readonly #hits: CoverageHits = {};
+
+  #coveragePath?: string;
+  #hitsPath?: string;
+
+  #getCoveragePath(): string {
+    if (this.#coveragePath === undefined) {
+      const coveragePath = process.env.HARDHAT_COVERAGE_PATH;
+      assertHardhatInvariant(
+        coveragePath !== undefined,
+        "HARDHAT_COVERAGE_PATH was not set",
+      );
```

Comment:
This is to be set in the hre hook handler. One alternative would be to operate inside a well-known directory, like root level `/coverage` directory for example, and just make sure that it's cleared every time we start a new coverage collection.

Reply:
I did explore this idea a bit further and I think it is clearer than using the temporary folder. See https://github.com/NomicFoundation/hardhat/compare/coverage-6581...coverage-6581-alt

The only question is, what directory should we use. In that exploration, I added a new config path `hre.config.paths.coverage` which points to `./coverage` by default.